tests/stub/driver_parameters/test_bookmark_manager.py

```diff
@@ -0,0 +1,598 @@
+            uri, auth,
+            bookmark_manager_config=Neo4jBookmarkManagerConfig()
+        )
+
+        s1 = self._driver.session("w")
+        s1.run("QUERY1").consume()
+        s1.close()
+
+        s2 = self._driver.session("w")
+        s2.run("QUERY2").consume()
+        s2.close()
+
+        s3 = self._driver.session("w")
+        s3.run("QUERY3").consume()
+        s3.close()
+
+        run_requests = self._server.get_requests("RUN")
+
+        self.assertEqual(len(run_requests), 3)
+        self.assert_run(run_requests[0])
+        self.assert_run(
+            run_requests[1],
+            bookmarks=["bm1"]
+        )
+        self.assert_run(
+            run_requests[2],
+            bookmarks=["bm2"]
+        )
+
+    def test_should_keep_track_of_tx_in_sequence(self):
+        self._start_server(self._router, "router_with_db_name.script")
+        self._start_server(self._server, "transaction_chaining.script")
+
+        uri = "neo4j://%s" % self._router.address
+        auth = types.AuthorizationToken("basic", principal="neo4j",
+                                        credentials="pass")
+        self._driver = Driver(
+            self._backend,
+            uri, auth,
+            bookmark_manager_config=Neo4jBookmarkManagerConfig()
+        )
+
+        s1 = self._driver.session("w")
+        tx1 = s1.begin_transaction({"order": "1st"})
+        tx1.run("RETURN 1 as n").consume()
+        tx1.commit()
+        s1.close()
+
+        s2 = self._driver.session("w")
+        tx2 = s2.begin_transaction({"order": "2nd"})
+        tx2.run("RETURN 1 as n").consume()
+        tx2.commit()
+        s2.close()
+
+        s3 = self._driver.session("w")
+        tx3 = s3.begin_transaction({"order": "2nd"})
+        tx3.run("RETURN 1 as n").consume()
+        tx3.commit()
+        s3.close()
+
+        begin_requests = self._server.get_requests("BEGIN")
+
+        self.assertEqual(len(begin_requests), 3)
+        self.assert_begin(begin_requests[0])
+        self.assert_begin(
+            begin_requests[1],
+            bookmarks=["bm1"]
+        )
+        self.assert_begin(
+            begin_requests[2],
+            bookmarks=["bm2"]
+        )
+
+    def test_should_not_replace_bookmarks_by_empty_bookmarks(self):
+        self._start_server(self._router, "router_with_db_name.script")
+        self._start_server(self._server, "transaction_chaining.script")
+
+        uri = "neo4j://%s" % self._router.address
+        auth = types.AuthorizationToken("basic", principal="neo4j",
+                                        credentials="pass")
+        self._driver = Driver(
+            self._backend,
+            uri, auth,
+            bookmark_manager_config=Neo4jBookmarkManagerConfig()
+        )
+
+        s1 = self._driver.session("w")
+        tx1 = s1.begin_transaction({"order": "1st"})
+        tx1.run("RETURN 1 as n").consume()
+        tx1.commit()
+        s1.close()
+
+        s2 = self._driver.session("r")
```

Comment:
I think it's not ideal to use the access mode to assert the the BMM ignores empty bookmark lists. The driver could also (erroneously) pass this test but never updating the BMM for any read-access queryies.